File: observalog-node/src/fields.ts
export type F = Record<string, unknown>;

export class Err {
    constructor(
        public readonly kind: string,
        public readonly code: string,
        public readonly message: string,
        public readonly retryable: boolean,
    ) {}
}

export enum Outcome {
    Success = 'success',
    Failure = 'failure',
    Partial = 'partial',
    Pending = 'pending',
}

export enum Level {
    Debug = 0,
    Info  = 1,
    Warn  = 2,
    Error = 3,
}

File: observalog-go/fields.go
package log

// F is the developer-supplied context map attached to every log call.
type F map[string]interface{}

// Err carries typed error metadata. Never pass a raw error string.
type Err struct {
	Kind      string // error class: "RateLimitExceeded", "NetworkError"
	Code      string // machine-readable constant: "PROVIDER_QUOTA_EXCEEDED"
	Message   string // err.Error() value
	Retryable bool
}

// Outcome represents the result of a decision point or operation exit.
type Outcome string

const (
	Success Outcome = "success"
	Failure Outcome = "failure"
	Partial Outcome = "partial"
	Pending Outcome = "pending"
)

// Level controls which log calls are emitted vs filtered.
// Ordered: Debug < Info < Warn < Error — comparable with <.
type Level uint8

const (
	LevelDebug Level = iota
	LevelInfo
	LevelWarn
	LevelError
)

func (l Level) String() string {
	switch l {
	case LevelDebug:
		return "debug"
	case LevelInfo:
		return "info"
	case LevelWarn:
		return "warn"
	case LevelError:
		return "error"
	default:
		return "unknown"
	}
}
